staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,519 @@
+		}
+
+		// Record the conflict in the conflicts cache.
+		if err = v.recordConflict(potentialRival, candidate); err != nil {
+			return conflictsMsgs, conflictFound, err
+		}
+	}
+
+	return conflictsMsgs, conflictFound, nil
+}
+
+func (v *Validator) updateSubnetValidity(s *netv1a1.Subnet, validated bool, validationErrors []string) error {
+	sCopy := s.DeepCopy()
+
+	sCopy.Status.Validated = validated
+	sCopy.Status.Errors.Validation = validationErrors
+
+	_, err := v.netIfc.Subnets(sCopy.Namespace).Update(sCopy)
+	switch {
+	case err == nil:
+		nsn := k8stypes.NamespacedName{
+			Namespace: s.Namespace,
+			Name:      s.Name,
+		}
+		v.updateStaleRV(nsn, s.ResourceVersion)
+	case doNotRetryUpdate(err):
+		glog.Errorf("failed to update subnet from %#+v to %#+v: %s. There will be no retry because of the nature of the error", s, sCopy, err.Error())
+	case !doNotRetryUpdate(err):
+		return fmt.Errorf("failed to update subnet from %#+v to %#+v: %s", s, sCopy, err.Error())
+	}
+
+	return nil
+}
+
+func (v *Validator) recordConflict(enroller, enrollee *subnetData) error {
+	v.conflictsMutex.Lock()
+	defer v.conflictsMutex.Unlock()
+
+	c := v.conflicts[enroller.namespacedName]
+
+	if c == nil {
+		return fmt.Errorf("registration of %s as a rival of %s failed: %s's conflicts cache not found", enrollee.namespacedName, enroller.namespacedName, enroller.namespacedName)
+	}
+
+	if enroller.differs(c.ownerData) {
+		// If we're here the version of the enroller recorded in its conflicts
+		// cache does not match the version of the enroller we got with the live
+		// list to the API server: one of the two is stale. Return an error so
+		// that the caller can wait a little bit and retry (hopefully the
+		// version skew has resolved by then).
+		return fmt.Errorf("registration of %s as a rival of %s failed: mismatch between %s's live data (%#+v) and conflicts cache data (%#+v)", enrollee.namespacedName, enroller.namespacedName, enroller.namespacedName, enroller, c.ownerData)
+	}
+
+	c.rivals = append(c.rivals, enrollee.namespacedName)
+	return nil
+}
+
+func (v *Validator) updateStaleRV(s k8stypes.NamespacedName, rv string) {
+	v.staleRVsMutex.Lock()
+	defer v.staleRVsMutex.Unlock()
+
+	v.staleRVs[s] = rv
+}
+
+func doNotRetryUpdate(e error) bool {
+	return k8serrors.IsUnauthorized(e) ||
+		k8serrors.IsBadRequest(e) ||
+		k8serrors.IsForbidden(e) ||
+		k8serrors.IsNotAcceptable(e) ||
+		k8serrors.IsUnsupportedMediaType(e) ||
+		k8serrors.IsMethodNotSupported(e) ||
+		k8serrors.IsInvalid(e) ||
+		k8serrors.IsGone(e)
+}
+
+// subnetData is the subset of the fields of a subnet relevant to determine
+// whether there are conflicts with other subnets.
+type subnetData struct {
+	namespacedName    k8stypes.NamespacedName
+	vni, baseU, lastU uint32
+}
+
+func ipv4ToUint32(ip net.IP) uint32 {
+	v4 := ip.To4()
+	return uint32(v4[0])<<24 + uint32(v4[1])<<16 + uint32(v4[2])<<8 + uint32(v4[3])
```

Comment:
why not do it like the IPAM controller does?

Reply:
What are you referring to exactly? The code in `ipv4ToUint32` is identical to the one in the IPAM. One change I vote for is factoring out all conversion functions in a separate package and have the IPAM, the subnets validator and the subnet strategy code import said package.

Reply:
Oh, sorry, I was remembering wrong.  I was thinking of using `binary.BigEndian`, like this:

```
func MakeUint32FromIPv4(ip []byte) uint32 {
	var iv uint32
	if len(ip) > 4 {
		iv = binary.BigEndian.Uint32(ip[len(ip)-4:])
	} else {
		iv = binary.BigEndian.Uint32(ip)
	}
	return iv
}

func MakeIPv4FromUint32(n uint32) net.IP {
	ip := make(net.IP, 4)
	binary.BigEndian.PutUint32(ip, n)
	return ip
}
```

Yes, putting stuff like this in a common place would be good.

Reply:
392c245 changes the implementation of the aforementioned functions to what you suggest and puts them in a common place.